ts/components/messages/paginated/MessageList/useMessagesSelection.tsx

```diff
@@ -0,0 +1,64 @@
+import React from "react";
+import { useItemsSelection } from "../../../../utils/hooks/useItemsSelection";
+import { UIMessage } from "../../../../store/reducers/entities/messages/types";
+import ListSelectionBar from "../../../ListSelectionBar";
+
+export const useMessagesSelection = (
+  allMessages: ReadonlyArray<UIMessage>,
+  navigateToMessageDetail: (message: UIMessage) => void,
+  primaryActionText: string,
+  primaryAction: (selectedItems: ReadonlyArray<UIMessage>) => void
+) => {
+  const { selectedItems, toggleItem, setAllItems, resetSelection } =
+    useItemsSelection();
+
+  const isSelecting = selectedItems.isSome();
+  const selectedItemsCount = selectedItems.toUndefined()?.size ?? 0;
+  const allItemsCount = allMessages.length;
+
+  const onPressItem = (message: UIMessage) => {
+    if (selectedItems.isSome()) {
+      toggleItem(message.id);
+    } else {
+      navigateToMessageDetail(message);
+    }
+  };
+
+  const onLongPressItem = (id: string) => {
+    toggleItem(id);
+  };
+
+  const onToggleAllSelection = () => {
+    if (selectedItemsCount === allItemsCount) {
+      setAllItems([]);
+    } else {
+      setAllItems(allMessages.map(_ => _.id));
+    }
+  };
+
+  const MessagesSelectionBar = () =>
```

Comment:
⛏️ **nitpick**: I understand why you are doing this, but I have a feeling that having a hook returning a component may lead to very difficult debugging situation.
I see two reasons for this:

1. hooks are stateful functions managed by React under the hood, and the component may not follow the update cycles we expect from its position in the tree
2. it mixes up logic with presentation, should we need a slightly different requirement for `ListSelectionBar` we would likely start injecting styles and ad-hoc props into the hook

I suggest we keep the two separated. It's ok to repeat some markup here and there :-)

Reply:
@pietro909 ok, I can revert it. Thanks for the tip.